librepo/metadata_downloader.c

```diff
@@ -465,6 +462,96 @@ restore_handle_callbacks(GSList *targets, GSList *handle_callbacks_backups)
     g_slist_free(handle_callbacks_backups);
 }
 
+static void
+create_metalink_or_mirrorlist_download_targets(GSList *targets, GSList **download_targets)
+{
+    for (GSList *elem = targets; elem; elem = g_slist_next(elem)) {
+        LrMetadataTarget *target = elem->data;
+        LrHandle *handle;
+        // handle is required
+        assert(target->handle);
+        handle = target->handle;
+
+        if (!handle->metalinkurl && !handle->mirrorlisturl) {
+            // No metalink or mirrorlist to download
+            continue;
+        }
+
+        if (handle->offline || handle->local) {
+            // Handle is configured not to download
+            continue;
+        }
+
+        int fd = lr_gettmpfile();
+        if (fd < 0) {
+            lr_metadatatarget_append_error(target, "Cannot create a temporary file");
+            continue;
+        }
+
+        _cleanup_free_ gchar *url;
+
+        if (handle->metalinkurl) {
+            url = lr_prepend_url_protocol(handle->metalinkurl);
+        } else if (handle->mirrorlisturl) {
+            url = lr_prepend_url_protocol(handle->mirrorlisturl);
+        } else {
+            // Creating targets only for metalinks or mirrorlists
+            assert(false);
+        }
+        handle->onetimeflag_apply = TRUE;
```

Comment:
There might be an issue here... CI test `dnf/countme.feature:124  Countme flag is not sent repeatedly on retries` is failing (ran locally with https://github.com/rpm-software-management/dnf5/pull/2210).

Reply:
Oh, I have missed that test.
I found out the problem is not here but in how many retries are made if the mirror is failing. I have updated the PR with fix and a comment.
I have also noticed the original (non-parallel) code path is downloading both mirrorlist and metalink if they are both available so I made additional changes to do that also in the parallel downloads.